src/stream/callback.rs

```diff
@@ -218,75 +228,160 @@ impl Default for Callback {
-                    if let Ok(mut counter) = counter.try_lock() {
-                        *counter += 1;
-                        if *counter > 100 {
-                            *counter = 0;
-                            let arguments = CallbackArguments { current_chunk: data };
-                            closure(arguments)
+                    match data {
+                        InternalSignals::Value(data) => {
+                            if let Ok(mut counter) = counter.try_lock() {
+                                *counter += 1;
+                                if *counter > 100 {
+                                    *counter = 0;
+                                    let arguments = CallbackArguments { current_chunk: data };
+                                    closure(arguments)
+                                }
+                            }
                         }
+                        InternalSignals::Finished => break,
                     }
                 }
             }
             OnProgressType::SlowAsyncClosure(closure) => {
                 while let Some(data) = receiver.recv().await {
-                    if let Ok(mut counter) = counter.try_lock() {
-                        *counter += 1;
-                        if *counter > 100 {
-                            *counter = 0;
-                            let arguments = CallbackArguments { current_chunk: data };
-                            closure(arguments).await
+                    match data {
+                        InternalSignals::Value(data) => {
+                            if let Ok(mut counter) = counter.try_lock() {
+                                *counter += 1;
+                                if *counter > 100 {
+                                    *counter = 0;
+                                    let arguments = CallbackArguments { current_chunk: data };
+                                    closure(arguments).await
+                                }
+                            }
                         }
+                        InternalSignals::Finished => break,
                     }
                 }
             }
             OnProgressType::SlowChannel(sender, cancel_on_close) => {
                 while let Some(data) = receiver.recv().await {
-                    if let Ok(mut counter) = counter.try_lock() {
-                        *counter += 1;
-                        if *counter > 100 {
-                            *counter = 0;
-                            let arguments = CallbackArguments { current_chunk: data };
-                            if sender.send(arguments).await.is_err() && cancel_on_close {
-                                receiver.close()
+                    match data {
+                        InternalSignals::Value(data) => {
+                            if let Ok(mut counter) = counter.try_lock() {
+                                *counter += 1;
+                                if *counter > 100 {
```

Comment:
Is this really correct? In `Stream::write_stream_to_file`, the counter is increased by `chunk.len()`, so if I'm not mistaken, a counter == 100 would mean 100 bytes, not 1 MB, right?

Reply:
As mentioned earlier, one chunk is ~10kb in size, varying between 1k and 16k. All of these methods are triggered per downloaded chunk.

Reply:
The counter only counts the number of messages and not their size. I'm currently refactoring this to sense the current progress and trigger once for every MB downloaded.

Reply:
> As mentioned earlier, one chunk is ~10kb in size, varying between 1k and 16k. All of these methods are triggered per downloaded chunk.

My bad, I mixed this `counter` and the `counter` form `write_stream_to_file` up.